File: web/src/timeline/edits.ts
/**
 * Pure timeline edit operations — the TS mirror of the server's
 * `cutroom.timeline.edits`. Every function takes a {@link Timeline} and returns
 * a NEW Timeline; the input is never mutated (each op deep-clones first). Clips
 * are identified by `id`; an unknown id (or a kind the op doesn't apply to)
 * yields an unchanged clone so callers can treat every op as total.
 *
 * Rational (frame) time is preserved: a clip's timeline position
 * (`start`/`duration`) stays independent of the source slice it shows
 * (`source_start`/`source_end`), which is what makes slip, ripple trims,
 * freeze-tail-as-an-edit and non-destructive splits expressible.
 *
 * No `validate()` exists on the client (the server is the source of truth); we
 * keep the derived convenience fields (`total_frames`, `duration_seconds`) in
 * sync after every edit so the returned object is internally consistent.
 */

import {
  Clip,
  Timeline,
  clipEnd,
  effectiveSourceEnd,
  framesToSeconds,
  totalFrames,
} from "./model";

const MEDIA_KINDS: ReadonlySet<Clip["kind"]> = new Set(["video", "image", "audio"]);

const newId = (prefix = "c_"): string =>
  prefix + Math.random().toString(16).slice(2, 10) + Math.random().toString(16).slice(2, 6);

/** Deep clone via JSON round-trip so the input timeline is never touched. */
const clone = (tl: Timeline): Timeline => JSON.parse(JSON.stringify(tl)) as Timeline;

const findClip = (tl: Timeline, clipId: string): Clip | undefined =>
  tl.clips.find((c) => c.id === clipId);

const isMedia = (c: Clip): boolean => MEDIA_KINDS.has(c.kind) && !!c.source;

/** Recompute the server-derived convenience fields, then return the timeline. */
const finalize = (tl: Timeline): Timeline => {
  tl.total_frames = totalFrames(tl);
  tl.duration_seconds = Math.round(framesToSeconds(tl.total_frames, tl.fps) * 1000) / 1000;
  return tl;
};

const sourceStartOf = (c: Clip): number => c.source_start ?? 0;

// --- source-window edits (position & duration fixed) ----------------------

/**
 * Slip a clip's source in/out window by `deltaFrames`, keeping timeline start &
 * duration fixed. Clamped to the available media via the clip's handles; excess
 * is clamped rather than raised.
 */
export function slip(tl: Timeline, clipId: string, deltaFrames: number): Timeline {
  const out = clone(tl);
  const c = findClip(out, clipId);
  if (!c || !isMedia(c)) return finalize(out);

  const inPt = sourceStartOf(c);
  const outPt = effectiveSourceEnd(c);

  let delta = deltaFrames;
  delta = Math.max(delta, -inPt); // can't slip earlier than the head
  if (c.source_duration != null) {
    delta = Math.min(delta, c.source_duration - outPt); // can't slip past the tail
  }

  c.source_start = inPt + delta;
  c.source_end = outPt + delta;
  return finalize(out);
}

/**
 * Set an explicit source in/out point. When the range length equals the clip
 * duration this is a pure slip; when it differs, `start` is pinned and
 * `duration` is set to the window length (a retime-free trim). Both branches
 * collapse to: start unchanged, window = [sourceStart, sourceEnd), duration =
 * window length. In-point is clamped to >= 0.
 */
export function set_source_range(
  tl: Timeline,
  clipId: string,
  sourceStart: number,
  sourceEnd: number,
): Timeline {
  const out = clone(tl);
  const c = findClip(out, clipId);
  if (!c || !isMedia(c)) return finalize(out);

  const ss = Math.max(0, Math.trunc(sourceStart));
  const se = Math.trunc(sourceEnd);
  c.source_start = ss;
  c.source_end = se;
  c.duration = se - ss;
  return finalize(out);
}

// --- ripple trims ----------------------------------------------------------

/**
 * Ripple-trim the clip's IN edge. `deltaFrames > 0` extends the head (earlier
 * in-point, longer clip); `< 0` shortens it. `start` and the out-point stay
 * fixed; `duration` changes by the applied delta and every later clip on the
 * same track shifts by that net delta. Bounded by `headHandle` (extend) and
 * duration/window >= 1 (shorten); excess is clamped.
 */
export function ripple_trim_start(
  tl: Timeline,
  clipId: string,
  deltaFrames: number,
): Timeline {
  const out = clone(tl);
  const c = findClip(out, clipId);
  if (!c || !isMedia(c)) return finalize(out);

  const origEnd = clipEnd(c);
  const inPt = sourceStartOf(c);
  const window = effectiveSourceEnd(c) - inPt;

  let applied: number;
  if (deltaFrames >= 0) {
    applied = Math.min(deltaFrames, inPt); // headHandle === inPt
  } else {
    const maxShorten = Math.min(c.duration, window) - 1;
    applied = Math.max(deltaFrames, -maxShorten);
  }

  // in-point moves by -applied; out-point stays fixed (source_end untouched).
  c.source_start = inPt - applied;
  c.duration = c.duration + applied;
  rippleAfter(out, c, origEnd, applied);
  return finalize(out);
}

/**
 * Ripple-trim the clip's OUT edge. `deltaFrames > 0` extends the tail; `< 0`
 * shortens it. `start` and the in-point stay fixed; `duration` changes by the
 * applied delta and every later clip on the same track shifts by that net delta.
 * Bounded by `tailHandle` (extend, when source_duration is known) and
 * duration/window >= 1 (shorten); excess is clamped.
 */
export function ripple_trim_end(
  tl: Timeline,
  clipId: string,
  deltaFrames: number,
): Timeline {
  const out = clone(tl);
  const c = findClip(out, clipId);
  if (!c || !isMedia(c)) return finalize(out);

  const origEnd = clipEnd(c);
  const outPt = effectiveSourceEnd(c);
  const window = outPt - sourceStartOf(c);

  let applied: number;
  if (deltaFrames >= 0) {
    applied = c.source_duration != null ? Math.min(deltaFrames, c.source_duration - outPt) : deltaFrames;
  } else {
    const maxShorten = Math.min(c.duration, window) - 1;
    applied = Math.max(deltaFrames, -maxShorten);
  }

  c.source_end = outPt + applied;
  c.duration = c.duration + applied;
  rippleAfter(out, c, origEnd, applied);
  return finalize(out);
}

/** Shift every clip on the edited clip's track starting at/after `thresholdEnd`
 * by `shift` frames (in place, on the working clone). */
function rippleAfter(tl: Timeline, edited: Clip, thresholdEnd: number, shift: number): void {
  if (shift === 0) return;
  for (const c of tl.clips) {
    if (c.id !== edited.id && c.track_id === edited.track_id && c.start >= thresholdEnd) {
      c.start = Math.max(0, c.start + shift);
    }
  }
}

// --- structural edits ------------------------------------------------------

/**
 * Split a clip at absolute timeline frame `atFrame` into two clips sharing the
 * same source. The right clip's `source_start` advances by (atFrame - start) so
 * the source is continuous across the cut; both carry a shared
 * `cutroom.origin_id`. Returns unchanged if `atFrame` is on/outside the clip.
 */
export function split_clip(tl: Timeline, clipId: string, atFrame: number): Timeline {
  const out = clone(tl);
  const c = findClip(out, clipId);
  if (!c) return finalize(out);
  if (atFrame <= c.start || atFrame >= clipEnd(c)) return finalize(out);

  const offset = atFrame - c.start;
  const origin = (c.cutroom?.origin_id as string | undefined) ?? c.id;

  const second: Clip = {
    id: newId("c_"),
    track_id: c.track_id,
    kind: c.kind,
    start: atFrame,
    duration: c.duration - offset,
    source: c.source,
    source_start: c.source ? sourceStartOf(c) + offset : 0,
    source_end: c.source ? effectiveSourceEnd(c) : undefined,
    source_duration: c.source_duration,
    source_fps: c.source_fps,
    label: c.label,
    text: c.text,
    color: c.color,
    cutroom: { ...(c.cutroom ?? {}), origin_id: origin },
  };

  if (c.source) c.source_end = sourceStartOf(c) + offset;
  c.duration = offset;
  c.cutroom = { ...(c.cutroom ?? {}), origin_id: origin };

  const idx = out.clips.indexOf(c);
  out.clips.splice(idx + 1, 0, second);
  return finalize(out);
}

/** Reposition a clip to a new timeline `start` (no ripple). Clamped to >= 0. */
export function move_clip(tl: Timeline, clipId: string, newStart: number): Timeline {
  const out = clone(tl);
  const c = findClip(out, clipId);
  if (!c) return finalize(out);
  c.start = Math.max(0, Math.trunc(newStart));
  return finalize(out);
}

/**
 * Remove a clip. With `ripple=true` the gap is closed by shifting every later
 * clip on the same track left by the removed clip's duration.
 */
export function remove_clip(tl: Timeline, clipId: string, ripple = false): Timeline {
  const out = clone(tl);
  const c = findClip(out, clipId);
  if (!c) return finalize(out);

  const trackId = c.track_id;
  const removedEnd = clipEnd(c);
  const removedDur = c.duration;
  out.clips = out.clips.filter((x) => x.id !== clipId);

  if (ripple) {
    for (const o of out.clips) {
      if (o.track_id === trackId && o.start >= removedEnd) {
        o.start = Math.max(0, o.start - removedDur);
      }
    }
  }
  return finalize(out);
}

/**
 * The FIRST-SECOND LAW as a non-destructive edit: trim a VIDEO clip to its first
 * `liveFrames` frames (source_end = source_start + liveFrames, duration =
 * liveFrames). Only valid on video clips; any other kind is returned unchanged.
 * `liveFrames` outside [1, duration) is a no-op.
 */
export function freeze_tail_trim(tl: Timeline, clipId: string, liveFrames: number): Timeline {
  const out = clone(tl);
  const c = findClip(out, clipId);
  if (!c || c.kind !== "video") return finalize(out);

  const live = Math.trunc(liveFrames);
  if (live < 1 || live >= c.duration) return finalize(out);

  c.source_end = sourceStartOf(c) + live;
  c.duration = live;
  return finalize(out);
}

File: server/cutroom/timeline/edits.py
"""Pure timeline edit operations.

Every function here takes a :class:`Timeline` and returns a **new, validated**
Timeline — the input is never mutated. Purity is enforced structurally: each op
deep-copies the input first (``Timeline.from_dict(tl.to_dict())``), edits the
copy, then calls ``.validate()`` before returning, so an operation can only ever
hand back a timeline the model considers legal.

Rational (frame) time is preserved throughout — a clip's timeline position
(``start``/``duration``) is kept independent of the slice of source it shows
(``source_start``/``source_end``), which is exactly what makes slip, ripple
trims, freeze-tail-as-an-edit and non-destructive splits expressible instead of
destructive.

Clips are always identified by ``id``. If the id is not found (or the op does not
apply to the clip's kind) the function returns an unchanged — but freshly copied
and re-validated — timeline rather than raising, so callers can treat every op
as total.
"""
from __future__ import annotations

from typing import Optional

from .model import MEDIA_KINDS, Clip, Timeline, new_id

__all__ = [
    "slip",
    "set_source_range",
    "ripple_trim_start",
    "ripple_trim_end",
    "split_clip",
    "move_clip",
    "remove_clip",
    "freeze_tail_trim",
]


# --- internal helpers -----------------------------------------------------


def _clone(tl: Timeline) -> Timeline:
    """Deep copy via the JSON round-trip so the input is never touched."""
    return Timeline.from_dict(tl.to_dict())


def _find(tl: Timeline, clip_id: str) -> Optional[Clip]:
    return next((c for c in tl.clips if c.id == clip_id), None)


def _is_media(c: Clip) -> bool:
    return c.kind in MEDIA_KINDS and bool(c.source)


# --- source-window edits (position & duration fixed) ----------------------


def slip(tl: Timeline, clip_id: str, delta_frames: int) -> Timeline:
    """Slip a clip's source in/out window by ``delta_frames``.

    The timeline ``start`` and ``duration`` stay put; only *which* frames of the
    source play changes: ``source_start`` and ``source_end`` both shift by delta.

    Clamped to the available media using the clip's handles — you cannot slip the
    in-point below 0 (``head_handle``) nor the out-point past ``source_duration``
    (``tail_handle``). If the requested delta would exceed the media it is clamped
    to the maximum possible slip and **no error is raised**.
    """
    out = _clone(tl)
    c = _find(out, clip_id)
    if c is None or not _is_media(c):
        return out.validate()

    in_pt = c.source_start
    out_pt = c.effective_source_end

    delta = delta_frames
    # lower bound: in_pt + delta >= 0  (can't slip earlier than the head)
    delta = max(delta, -in_pt)
    # upper bound: out_pt + delta <= source_duration (can't slip past the tail)
    if c.source_duration is not None:
        delta = min(delta, c.source_duration - out_pt)

    c.source_start = in_pt + delta
    c.source_end = out_pt + delta
    return out.validate()


def set_source_range(tl: Timeline, clip_id: str, source_start: int,
                     source_end: int) -> Timeline:
    """Set an explicit source in/out point.

    Design choice for the "slip vs. trim" ambiguity: a *pure slip* keeps the
    timeline duration and only redefines which source frames play — but that is
    only meaningful when the new range has the **same length** as the current
    clip duration. When the requested range is a different length, keeping the
    old duration would imply a retime, which is out of scope here. So this op
    treats the range as authoritative: it pins the timeline ``start`` and sets
    ``duration = source_end - source_start`` (a retime-free trim-in-place).

    Both branches therefore collapse to the same rule — ``start`` unchanged,
    window = ``[source_start, source_end)``, ``duration`` = the window length —
    and when the length happens to equal the old duration the result is exactly
    a slip. An out-of-media or empty range is rejected by ``validate()``
    (raising ``TimelineError``); the in-point is clamped to >= 0.
    """
    out = _clone(tl)
    c = _find(out, clip_id)
    if c is None or not _is_media(c):
        return out.validate()

    ss = max(0, int(source_start))
    se = int(source_end)
    c.source_start = ss
    c.source_end = se
    c.duration = se - ss
    return out.validate()


# --- ripple trims (edge moves + downstream shifts to stay contiguous) ------


def ripple_trim_start(tl: Timeline, clip_id: str, delta_frames: int) -> Timeline:
    """Ripple-trim the clip's IN edge.

    ``delta_frames > 0`` extends the head (reveals earlier source frames — the
    in-point moves *earlier*, the clip gets longer); ``delta_frames < 0`` shortens
    it (in-point moves later, clip gets shorter). The clip's timeline ``start``
    and its out-point stay fixed, ``duration`` changes by the applied delta, and
    every later clip on the same track shifts by that same net delta so the track
    stays gap/overlap-free.

    Bounded by the handles: extension is limited by ``head_handle`` (available
    source before the in-point); shortening keeps ``duration >= 1`` and the source
    window non-empty. Excess is clamped, not raised.
    """
    out = _clone(tl)
    c = _find(out, clip_id)
    if c is None or not _is_media(c):
        return out.validate()

    orig_end = c.end
    in_pt = c.source_start
    window = c.effective_source_end - in_pt

    if delta_frames >= 0:
        applied = min(delta_frames, in_pt)              # head_handle == in_pt
    else:
        max_shorten = min(c.duration, window) - 1
        applied = max(delta_frames, -max_shorten)

    # in-point moves by -applied; out-point stays fixed (source_end untouched):
    #  - explicit source_end -> unchanged
    #  - implicit source_end -> effective end = source_start + duration, and both
    #    change by (-applied)/(+applied), so the out-point is preserved too.
    c.source_start = in_pt - applied
    c.duration = c.duration + applied
    _ripple_after(out, c, orig_end, applied)
    return out.validate()


def ripple_trim_end(tl: Timeline, clip_id: str, delta_frames: int) -> Timeline:
    """Ripple-trim the clip's OUT edge.

    ``delta_frames > 0`` extends the tail (out-point moves later, clip longer);
    ``delta_frames < 0`` shortens it. The clip's ``start`` and in-point stay
    fixed, ``duration`` changes by the applied delta, and every later clip on the
    same track shifts by that net delta to stay contiguous.

    Bounded by the handles: extension is limited by ``tail_handle`` (available
    source after the out-point, when ``source_duration`` is known); shortening
    keeps ``duration >= 1`` and the window non-empty. Excess is clamped.
    """
    out = _clone(tl)
    c = _find(out, clip_id)
    if c is None or not _is_media(c):
        return out.validate()

    orig_end = c.end
    out_pt = c.effective_source_end
    window = out_pt - c.source_start

    if delta_frames >= 0:
        if c.source_duration is not None:
            applied = min(delta_frames, c.source_duration - out_pt)  # tail_handle
        else:
            applied = delta_frames
    else:
        max_shorten = min(c.duration, window) - 1
        applied = max(delta_frames, -max_shorten)

    c.source_end = out_pt + applied
    c.duration = c.duration + applied
    _ripple_after(out, c, orig_end, applied)
    return out.validate()


def _ripple_after(tl: Timeline, edited: Clip, threshold_end: int, shift: int) -> None:
    """Shift every clip on the edited clip's track that starts at/after
    ``threshold_end`` by ``shift`` frames (in place, on the working copy)."""
    if shift == 0:
        return
    for c in tl.clips:
        if c.id != edited.id and c.track_id == edited.track_id and c.start >= threshold_end:
            c.start = max(0, c.start + shift)


# --- structural edits ------------------------------------------------------


def split_clip(tl: Timeline, clip_id: str, at_frame: int) -> Timeline:
    """Split a clip at absolute timeline frame ``at_frame`` into two clips that
    share the same source.

    The left clip keeps the original ``start`` and is shortened to end at
    ``at_frame``; the right clip begins at ``at_frame`` and its ``source_start``
    advances by ``(at_frame - original.start)`` so the source runs continuously
    across the cut. Both carry the original lineage plus a shared
    ``cutroom['origin_id']`` linking them back to the pre-split clip.

    Returns the timeline unchanged if ``at_frame`` falls on or outside the clip's
    bounds (nothing to split).
    """
    out = _clone(tl)
    c = _find(out, clip_id)
    if c is None:
        return out.validate()
    if at_frame <= c.start or at_frame >= c.end:
        return out.validate()

    offset = at_frame - c.start
    origin = c.cutroom.get("origin_id") or c.id

    # Build the right half from the ORIGINAL clip values first (before mutating).
    second = Clip(
        id=new_id("c_"),
        track_id=c.track_id,
        kind=c.kind,
        start=at_frame,
        duration=c.duration - offset,
        source=c.source,
        source_start=(c.source_start + offset) if c.source else 0,
        source_end=c.effective_source_end if c.source else None,
        source_duration=c.source_duration,
        source_fps=c.source_fps,
        label=c.label,
        text=c.text,
        color=c.color,
        cutroom=dict(c.cutroom),
    )
    second.cutroom["origin_id"] = origin

    # Shrink the left half in place.
    if c.source:
        c.source_end = c.source_start + offset
    c.duration = offset
    c.cutroom["origin_id"] = origin

    out.clips.insert(out.clips.index(c) + 1, second)
    return out.validate()


def move_clip(tl: Timeline, clip_id: str, new_start: int) -> Timeline:
    """Reposition a clip to a new timeline ``start`` (no ripple — later clips are
    left where they are). ``start`` is clamped to >= 0."""
    out = _clone(tl)
    c = _find(out, clip_id)
    if c is None:
        return out.validate()
    c.start = max(0, int(new_start))
    return out.validate()


def remove_clip(tl: Timeline, clip_id: str, ripple: bool = False) -> Timeline:
    """Remove a clip. With ``ripple=True`` the gap is closed by shifting every
    later clip on the same track left by the removed clip's duration; otherwise
    the gap is left in place."""
    out = _clone(tl)
    c = _find(out, clip_id)
    if c is None:
        return out.validate()

    track_id = c.track_id
    removed_end = c.end
    removed_dur = c.duration
    out.clips = [x for x in out.clips if x.id != clip_id]

    if ripple:
        for o in out.clips:
            if o.track_id == track_id and o.start >= removed_end:
                o.start = max(0, o.start - removed_dur)
    return out.validate()


def freeze_tail_trim(tl: Timeline, clip_id: str, live_frames: int) -> Timeline:
    """The FIRST-SECOND LAW as a non-destructive edit.

    Trim a **video** clip down to its first ``live_frames`` frames:
    ``source_end = source_start + live_frames`` and ``duration = live_frames``.
    The held-tail portion (freezing the last live frame for the remainder) is a
    separate future concern — this op only performs the trim.

    Only valid on video clips; on any other kind the timeline is returned
    unchanged. ``live_frames`` outside ``[1, duration)`` is a no-op (there is
    nothing to trim), so the op never lengthens a clip or produces an empty one.
    """
    out = _clone(tl)
    c = _find(out, clip_id)
    if c is None or c.kind != "video":
        return out.validate()

    live = int(live_frames)
    if live < 1 or live >= c.duration:
        return out.validate()

    c.source_end = c.source_start + live
    c.duration = live
    return out.validate()
